pkg/tlsconfig/certconfig_test.go

```diff
@@ -0,0 +1,485 @@
+		cert, err := cl.GetCertificate(nil)
+		require.NoError(t, err)
+		require.NotNil(t, cert)
+		require.Equal(t, cl.Certificate(), cert)
+		require.NotEmpty(t, cert.Certificate, "expected at least 1 certificate")
+		require.NotNil(t, cl.Leaf())
+		x509Cert, err := x509.ParseCertificate(cert.Certificate[0])
+		require.NoError(t, err)
+		require.NotNil(t, x509Cert)
+		require.Equal(t, []string{DNSName}, x509Cert.DNSNames)
+		require.Equal(t, x509Cert, cl.Leaf())
+	}
+
+	{
+		// Create new certificate and reload
+		const DNSName2 = "data1-ultimate-form.influxdata.edge"
+		logs.TakeAll()
+		ss := selfsigned.NewSelfSignedCert(t, selfsigned.WithDNSName(DNSName2))
+		require.NoError(t, cl.Load(ss.CertPath, ss.KeyPath))
+
+		require.Equal(t, 3, logs.Len())
+		logLines := logs.TakeAll()
+		require.Equal(t, "Loading TLS certificate (start)", logLines[0].Message)
+		require.Equal(t, "Successfully loaded TLS certificate", logLines[1].Message)
+		require.Equal(t, "Loading TLS certificate (end)", logLines[2].Message)
+		for _, l := range logLines {
+			cm := l.ContextMap()
+			require.Equal(t, ss.CertPath, cm["cert"])
+			require.Equal(t, ss.KeyPath, cm["key"])
+		}
+
+		cp, kp := cl.Paths()
+		require.Equal(t, ss.CertPath, cp)
+		require.Equal(t, ss.KeyPath, kp)
+		require.NotNil(t, cl.Certificate())
+		cert, err := cl.GetCertificate(nil)
+		require.NoError(t, err)
+		require.NotNil(t, cert)
+		require.Equal(t, cl.Certificate(), cert)
+		require.NotEmpty(t, cert.Certificate, "expected at least 1 certificate")
+		require.NotNil(t, cl.Leaf())
+		x509Cert, err := x509.ParseCertificate(cert.Certificate[0])
+		require.NoError(t, err)
+		require.NotNil(t, x509Cert)
+		require.Equal(t, []string{DNSName2}, x509Cert.DNSNames)
+		require.Equal(t, x509Cert, cl.Leaf())
+	}
+}
+
+func TestTLSCertLoader_GoodCertPersists(t *testing.T) {
+	const DNSName = "data1.influxdata.edge"
+	ss := selfsigned.NewSelfSignedCert(t, selfsigned.WithDNSName(DNSName))
+
+	core, logs := observer.New(zapcore.InfoLevel)
+	logger := zap.New(core)
+
+	// Start cert loader
+	cl, err := NewTLSCertLoader(ss.CertPath, ss.KeyPath, WithLogger(logger))
+	require.NoError(t, err)
+	require.NotNil(t, cl)
+	defer func() {
+		require.NoError(t, cl.Close())
+	}()
+
+	var goodSerial big.Int
+	{
+		cp, kp := cl.Paths()
+		require.Equal(t, ss.CertPath, cp)
+		require.Equal(t, ss.KeyPath, kp)
+
+		cert, err := cl.GetCertificate(nil)
+		require.NoError(t, err)
+		require.NotNil(t, cert)
+		require.NotEmpty(t, cert.Certificate)
+		x509Cert, err := x509.ParseCertificate(cert.Certificate[0])
+		require.NoError(t, err)
+		require.NotNil(t, x509Cert)
+		require.NotNil(t, x509Cert.SerialNumber)
+		goodSerial = *x509Cert.SerialNumber
+
+		// Logs for happy case are checked in another test, just dump them here.
+		logs.TakeAll()
+	}
+
+	{
+		// Create and load bad cert / key (empty files)
+		tmpdir := t.TempDir()
+		emptyFile, err := os.CreateTemp(tmpdir, "badcert-*.pem")
+		emptyPath := emptyFile.Name()
```

Comment:
Move this below the `NoError` check to avoid inadvertent `nil` pointer panics in the test on file creation failure.

Reply:
Will be fixed in next commit.